### PR TITLE
An SMT2 frontend plus solver binary

### PR DESCRIPTION
**Pull request by kroening**

### CODE REVIEW

**Comment by kroening** [comment]
Note that the actual solver doesn't link because of a pending dependency cleanup.

Will add some tests for the solver binary; suggestions for a sensible choice (not all of SMT-LIB) are welcome.

File: src/solvers/smt2/smt2_solver.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
+#include <fstream>
```
Comment (tautschnig) [inline src/solvers/smt2/smt2_solver.cpp:9]: No copyright information.

File: src/solvers/Makefile [hunk 2]
```diff
@@ -215,7 +215,7 @@ INCLUDES += -I .. \
 
 CLEANFILES = solvers$(LIBEXT)
 
-all: solvers$(LIBEXT)
+all: solvers$(LIBEXT) smt2_solver$(EXEEXT)
```
Comment (tautschnig) [inline src/solvers/Makefile]: The clean target, .gitignore, and CMake will need updates.

**Review by tautschnig — CHANGES_REQUESTED** [review]
1. Placing this new front-end in solvers/ requires a fair bit of further work (see also comments), and the way it is set up it will always be built when anyone depends on solvers, which seems rather unnecessary.
2. When refactoring, please add comments (at least at the level of classes).

File: src/solvers/smt2/smt2_solver.cpp (hunk 1 above)
Comment (kroening) [reply src/solvers/smt2/smt2_solver.cpp:9]: Done

**Comment by kroening** [comment]
yes, I am contemplating to not make the smt2_solver binary a default target; however, this would be the first such instance.

**Comment by kroening** [comment]
The cleanup is taking too long, will aim at getting this merged before.

**Comment by tautschnig** [comment]
> The cleanup is taking too long, will aim at getting this merged before.

The cleanup of what? And which PR should be merged before which?

**Comment by kroening** [comment]
A cleanup of the dependencies inside solver/; removing these enables the build of the smt2_solver binary without linking to goto-programs, langapi, and java_bytecode.

**Comment by polgreen** [comment]
There is something wrong with the let parser. The smt benchmarks in this folder fail: 
https://clc-gitlab.cs.uiowa.edu:2443/SMT-LIB-benchmarks/QF_BV/tree/master/2017-BuchwaldFried

File: .gitignore [hunk 3]
```diff
@@ -104,6 +104,7 @@ src/goto-instrument/goto-instrument.exe
 src/jbmc/jbmc
 src/musketeer/musketeer
 src/musketeer/musketeer.exe
+src/solvers/smt2/smt2_solver
```
Comment (tautschnig) [inline .gitignore:107]: Even though a later commit disables this: the .exe variant should be added now.

File: regression/Makefile [hunk 4]
```diff
@@ -16,6 +16,7 @@ DIRS = ansi-c \
        invariants \
        strings \
        jbmc-strings \
+#       smt2_solver \
```
Comment (tautschnig) [inline regression/Makefile]: What's the effect of having a commented-out line ending in a `\` in a block of lines that end in `\`? (It certainly confuses GitHub's syntax highlighting, but possibly also `make`.)

File: src/solvers/flattening/boolbv.cpp [hunk 5]
```diff
@@ -463,6 +459,15 @@ literalt boolbvt::convert_rest(const exprt &expr)
     return convert_quantifier(expr);
   else if(expr.id()==ID_exists)
     return convert_quantifier(expr);
+  else if(expr.id()==ID_let)
+  {
+    bvt bv=convert_let(to_let_expr(expr));
+
+    if(bv.size()!=1)
+      throw "convert_let returned non-bool bitvector";
```
Comment (tautschnig) [inline src/solvers/flattening/boolbv.cpp]: This should be an `INVARIANT`. Or, really, why does it not return a `literalt` instead?!

File: src/solvers/flattening/boolbv_let.cpp [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  const bvt value_bv=convert_bv(expr.value());
+
+  // We expect the bindings to be unique, and
+  // thus, these can go straight into the symbol to literal map.
+  // This property also allows us to cache any subexpressions.
+  const irep_idt &id=expr.symbol().get_identifier();
+
+  map.set_literals(id, expr.symbol().type(), value_bv);
+  bvt result_bv=convert_bv(expr.where());
+  map.erase_literals(id, expr.symbol().type());
```
Comment (tautschnig) [inline src/solvers/flattening/boolbv_let.cpp]: Why are the first created/set and then erased?!

File: src/solvers/flattening/boolbv_let.cpp [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+#include "boolbv.h"
+
+#include <util/std_expr.h>
+#include <util/std_types.h>
+
+bvt boolbvt::convert_let(const let_exprt &expr)
+{
+  const bvt value_bv=convert_bv(expr.value());
+
+  // We expect the bindings to be unique, and
```
Comment (tautschnig) [inline src/solvers/flattening/boolbv_let.cpp]: Bindings between ... what?

File: src/solvers/flattening/boolbv_map.cpp [hunk 8]
```diff
@@ -146,3 +146,13 @@ void boolbv_mapt::set_literals(
...
 }
+
+void boolbv_mapt::erase_literals(
+  const irep_idt &identifier,
+  const typet &type)
+{
+  map_entryt &map_entry=get_map_entry(identifier, type);
+
+  for(auto &l : map_entry.literal_map)
+    l.is_set=false;
```
Comment (tautschnig) [inline src/solvers/flattening/boolbv_map.cpp]: This does not seem to "erase" literals?

File: src/solvers/prop/prop_conv.cpp [hunk 9]
```diff
@@ -327,8 +327,23 @@ literalt prop_conv_solvert::convert_bool(const exprt &expr)
...
-    throw "let is todo";
+    const let_exprt &let_expr=to_let_expr(expr);
+
+    // first check whether this is all boolean
+    if(let_expr.value().type().id()==ID_bool &&
+       let_expr.where().type().id()==ID_bool)
+    {
+      literalt value=convert(let_expr.value());
+
+      // We expect the bindings to be unique, and
```
Comment (tautschnig) [inline src/solvers/prop/prop_conv.cpp]: Bindings between what?

File: src/solvers/Makefile [hunk 10]
```diff
@@ -217,7 +217,7 @@ INCLUDES += -I .. \
 
 CLEANFILES = solvers$(LIBEXT) smt2_solver$(EXEEXT)
 
-all: solvers$(LIBEXT) smt2_solver$(EXEEXT)
+all: solvers$(LIBEXT) # smt2_solver$(EXEEXT)
```
Comment (tautschnig) [inline src/solvers/Makefile]: The commit message does not say "why" it was disabled, and in particular this does not explain why earlier commits first enabled/added it while now it is being disabled.

**Comment by kroening** [comment]
The 2017-BuchwaldFried benchmarks now parse (they need back-end support for rol, but that's orthogonal).

File: regression/Makefile (hunk 4 above)
Comment (kroening) [reply regression/Makefile]: Removed

File: .gitignore (hunk 3 above)
Comment (kroening) [reply .gitignore:107]: done

File: src/solvers/flattening/boolbv.cpp (hunk 5 above)
Comment (kroening) [reply src/solvers/flattening/boolbv.cpp]: There are both cases, i.e., one where let is used in a boolean context and one where it produces bit-vectors. Clarified the error message.

File: src/solvers/flattening/boolbv_let.cpp (hunk 6 above)
Comment (kroening) [reply src/solvers/flattening/boolbv_let.cpp]: Added a comment. The symbol is meant to be visible during the recursive call only.

File: src/solvers/flattening/boolbv_let.cpp (hunk 7 above)
Comment (kroening) [reply src/solvers/flattening/boolbv_let.cpp]: Re-worded comment

File: src/solvers/flattening/boolbv_map.cpp (hunk 8 above)
Comment (kroening) [reply src/solvers/flattening/boolbv_map.cpp]: ok, it does now

File: src/solvers/prop/prop_conv.cpp (hunk 9 above)
Comment (kroening) [reply src/solvers/prop/prop_conv.cpp]: Same as above, reworded

File: src/solvers/Makefile (hunk 10 above)
Comment (kroening) [reply src/solvers/Makefile]: reworded

**Comment by kroening** [comment]
The Travis failure is spurious, of course.

**Comment by peterschrammel** [comment]
@kroening, cmake builds pass now.

**Review by tautschnig — APPROVED** [review]
I am not happy that a command-line tool now sits in the solvers/ directory, and is built every time, but this is a minor concern. Things look ok otherwise.

**Review by peterschrammel — APPROVED** [review]

**Review by chrisr-diffblue — APPROVED** [review]
Not entirely sure why the solvers CMakeLists.txt needed to have CXX_STANDARD set rather than inheriting from the top-level CMakeLists properties, but it does appear it's needed. Other than that, I'm happy with the parts the parts for which I'm a code owner.

**Pull request merged**